src/transactions/SetTrustLineFlagsOpFrame.cpp

```diff
@@ -0,0 +1,200 @@
+    {
+        innerResult().code(SET_TRUST_LINE_FLAGS_MALFORMED);
+        return false;
+    }
+
+    if (!isAssetValid(mSetTrustLineFlags.asset))
+    {
+        innerResult().code(SET_TRUST_LINE_FLAGS_MALFORMED);
+        return false;
+    }
+
+    if (!(getSourceID() == getIssuer(mSetTrustLineFlags.asset)))
+    {
+        innerResult().code(SET_TRUST_LINE_FLAGS_MALFORMED);
+        return false;
+    }
+
+    if (mSetTrustLineFlags.trustor == getSourceID())
+    {
+        innerResult().code(SET_TRUST_LINE_FLAGS_MALFORMED);
+        return false;
+    }
+
+    if (mSetTrustLineFlags.setFlags && mSetTrustLineFlags.clearFlags)
+    {
+        if ((*mSetTrustLineFlags.setFlags & *mSetTrustLineFlags.clearFlags) !=
+            0)
+        {
+            innerResult().code(SET_TRUST_LINE_FLAGS_MALFORMED);
+            return false;
+        }
+    }
+
+    if (mSetTrustLineFlags.setFlags)
+    {
+        // setFlags has the same restrictions as the trustline flags, so we can
+        // use trustLineFlagIsValid here
+        if (!trustLineFlagIsValid(*mSetTrustLineFlags.setFlags,
+                                  ledgerVersion) ||
+            (*mSetTrustLineFlags.setFlags & TRUSTLINE_CLAWBACK_ENABLED_FLAG) !=
+                0)
+        {
+            innerResult().code(SET_TRUST_LINE_FLAGS_MALFORMED);
+            return false;
+        }
+    }
+
+    if (mSetTrustLineFlags.clearFlags &&
+        !trustLineFlagMaskCheckIsValid(*mSetTrustLineFlags.clearFlags,
+                                       ledgerVersion))
+    {
+        innerResult().code(SET_TRUST_LINE_FLAGS_MALFORMED);
+        return false;
+    }
+
+    return true;
+}
+
+void
+SetTrustLineFlagsOpFrame::insertLedgerKeysToPrefetch(
+    UnorderedSet<LedgerKey>& keys) const
+{
+    if (mSetTrustLineFlags.asset.type() != ASSET_TYPE_NATIVE)
+    {
+        keys.emplace(
+            trustlineKey(mSetTrustLineFlags.trustor, mSetTrustLineFlags.asset));
+    }
+}
+
+bool
+SetTrustLineFlagsOpFrame::isAuthRevocationValid(AbstractLedgerTxn& ltx)
+{
+    LedgerTxn ltxSource(ltx); // ltxSource will be rolled back
+    auto header = ltxSource.loadHeader();
+    auto sourceAccountEntry = loadSourceAccount(ltxSource, header);
+
+    if (sourceAccountEntry.current().data.account().flags & AUTH_REVOCABLE_FLAG)
+    {
+        return true;
+    }
+
+    // This condition covers all three revoke scenarios
+    // 1. AUTHORIZED_FLAG -> AUTHORIZED_TO_MAINTAIN_LIABILITIES_FLAG
+    // 2. AUTHORIZED_FLAG -> 0
+    // 3. AUTHORIZED_TO_MAINTAIN_LIABILITIES_FLAG -> 0
+    //
+    // In all three scenarios above, clearingAnyAuth == true and
+    // settingAuthorized == false
+
+    bool clearingAnyAuth =
+        mSetTrustLineFlags.clearFlags &&
+        (*mSetTrustLineFlags.clearFlags & TRUSTLINE_AUTH_FLAGS) != 0;
+
+    bool settingAuthorized =
+        mSetTrustLineFlags.setFlags &&
+        (*mSetTrustLineFlags.setFlags & AUTHORIZED_FLAG) != 0;
```

Comment:
Note that the CAP is under-specified when attempting to transition from 0 to one of the authorized states when the issuer does not have `AUTH_REVOCABLE_FLAG`.

This returns false (and causes `SET_TRUST_LINE_FLAGS_CANT_REVOKE`) when trying to transition from `0 -> AUTHORIZED_TO_MAINTAIN_LIABILITIES_FLAG`).

Also, this returns true for `0 -> AUTHORIZED_FLAG` - yet `AllowTrustOp` fails in this situation with `ALLOW_TRUST_TRUST_NOT_REQUIRED` (in this PR). I think the CAP should be explicit on the fact we're aligning with CAP29 here.

In any case you want to to make sure that we have `iif` : `(revoke scenario) iif (clearingAnyAuth == true and settingAuthorized == false)`.

Reply:
1. For your point about `0 -> AUTHORIZED_TO_MAINTAIN_LIABILITIES_FLAG` - This function returns true, and there's a test case for this under the section `upgrade auth when not revocable`. 
2. I agree the CAP should be updated to mention that we're aligning with CAP29. I'll take care of this.
3. I believe the condition I have here is equivalent to the condition in your comment. Am I missing something?

Reply:
Oh I see - I got confused by the comment the logic itself is fine.

What this is really doing is enforce that either auth flags are not being unset or if they are, the `AUTHORIZED_FLAG` is set so that we end up with the most restrictive auth.

The comment talks about it backwards (revoke scenarios) instead of talking about which case is valid.